tools/codegen/api/ufunc.py

```diff
@@ -0,0 +1,175 @@
+from tools.codegen.model import (Argument, BaseTy, BaseType, ListType,
+                                 NativeFunctionsGroup, OptionalType,
+                                 SelfArgument, TensorOptionsArguments, Type,
+                                 DispatchKey)
+
+import tools.codegen.api.types as api_types
+from tools.codegen.api.types import (ArgName, BaseCType, Binding, ArrayRefCType,
+                                     ConstRefCType, OptionalCType, NamedCType,
+                                     tensorT, scalarT, intArrayRefT, dimnameListT,
+                                     optionalTensorRefT, optionalScalarRefT, CType,
+                                     BaseCppType, CType)
+
+from tools.codegen.api import cpp, structured
+from tools.codegen.utils import (mapMaybe, assert_never)
+
+from dataclasses import dataclass
+from typing import Union, List, Optional, cast, Iterable
+
+def kernel_name(g: NativeFunctionsGroup, dispatch_key: DispatchKey) -> str:
+    return f"ufunc_{g.functional.func.name.name}_{dispatch_key}"
+
+# Tensors are omitted (as they are stored in TensorIterator), everything else is
+# passed along  (technically, we can pass tensors along too, it just wastes
+# argument registers)
+#
+# NB: used for CPU only
+def dispatchstub_type(t: Type, *, binds: ArgName) -> Optional[NamedCType]:
+    r = cpp.valuetype_type(t, binds=binds)
+    if r is not None:
+        return r
+
+    if t == BaseType(BaseTy.Scalar):
+        return NamedCType(binds, ConstRefCType(BaseCType(scalarT)))
+    elif t == BaseType(BaseTy.Tensor):
+        return None
+    else:
+        raise AssertionError(f"unrecognized type {repr(t)}")
+
+def opmath_type(scalar_t: BaseCppType) -> BaseCppType:
+    if scalar_t == api_types.scalar_t:
+        return api_types.opmath_t
+    raise NotImplementedError
+
+# NB: Tensors in constructor are stored in opmath_t, not scalar_t
+# because Tensor in constructor = its a scalar tensor partially applied =
+# it can be higher precision and we want to compute in that higher precision
+#
+# NB: CUDA only
+def ufunctor_ctor_type(t: Type, *, binds: ArgName, scalar_t: BaseCppType) -> NamedCType:
+    r = cpp.valuetype_type(t, binds=binds)
+    if r is not None:
+        return r
+
+    if t == BaseType(BaseTy.Scalar):
+        return NamedCType(binds, BaseCType(opmath_type(scalar_t)))
+    elif t == BaseType(BaseTy.Tensor):
+        return NamedCType(binds, BaseCType(opmath_type(scalar_t)))
+    else:
+        raise AssertionError(f"unrecognized type {repr(t)}")
+
+# Only Tensors ever get passed directly to operator()
+#
+# NB: CUDA only
+# (Actually, this works for CPU too)
+def ufunctor_apply_type(t: Type, *, binds: ArgName, scalar_t: BaseCppType) -> NamedCType:
+    if t == BaseType(BaseTy.Tensor):
+        return NamedCType(binds, BaseCType(scalar_t))
```

Comment:
Seeing how there's only support for vanilla `Tensor` (and not, say, `c10::optional<Tensor>`)  made me look more into what "weird" pointwise ops we currently have.

This is more of a general question, that probably shouldn't affect this PR: But to what extent do you think we want to beef up the ufunc codegen over time to handle the weirder cases, versus try to enforce more uniformity across the pointwise ops that we have?

Random example that I dug up: `clamp.Tensor` is weird, because it accepts an optional `Tensor?` min and max. That's handled in the kernel ([here](https://github.com/pytorch/pytorch/blob/ec8d6777255821bed73b471eadddde068cd60c0b/aten/src/ATen/native/TensorCompare.cpp#L502)), which basically checks that both optional tensors are valid before calling into `DispatchStub`, in the other cases, it redispatches to a different overload.

I could imagine maybe beefing up the codegen to handle that case, but it might be pretty hard to generalize. Another option would be to write `clamp.Tensor` as a composite kernel, and avoid the problem for ufunc codegen (at a small perf cost).

Reply:
I think my idea was that ufunc codegen would only ever target anything that went through DispatchStub. So if you had a kernel that took an optional Tensor? and decided to route to one of several dispatch stubs depending on whether or not it was set or not, post ufunc codegen I'd have separate kernels for each dispatch stub and then a composite that did the routing (though, the composite would probably have an explicit autograd kernel). This is a little less efficient, since I have to redispatch through the dispatcher rather than DispatchStub, but there are other optimizations that probably would work.